### ISSUE TITLE
installing packages from git broken

### ISSUE THREAD

**Issue by evertlammerts** [kennethreitz/pipenv#204]
Fresh install:

```
$ pipenv --version
pipenv, version 3.3.6
```

Simple Pipfile:

```
$ cat Pipfile
[[source]]
url = "https://pypi.python.org/simple"
verify_ssl = true

[packages]
crayons = { git="https://github.com/kennethreitz/crayons.git", ref="v0.1.2" }
```

Creates what seems to me a valid lock file:

```
$ cat Pipfile.lock
{
    "_meta": {
        "hash": {
            "sha256": "39b2447b7dd6976804de490321577a2db0f4fc4327e0795a93436ea69afad9a8"
        },
        "requires": {},
        "sources": [
            {
                "verify_ssl": true,
                "url": "https://pypi.python.org/simple"
            }
        ]
    },
    "default": {
        "crayons from git+https://github.com/kennethreitz/crayons.git@v0.1.2#egg=crayons": {
            "hash": "sha256:a292b7ef29131345fd329ff5a94d1e2e59569be4dbd9fa8b99ff7114d72102a6",
            "version": "==0.1.2"
        },
        "colorama": {
            "hash": "sha256:a4c0f5bc358a62849653471e309dcc991223cf86abafbec17cd8f41327279e89",
            "version": "==0.3.7"
        }
    },
    "develop": {}
}
```

And now installing breaks:

```
$ pipenv install
No package provided, installing all dependencies.
Pipfile found at /[...]/Pipfile. Considering this to be the project home.
Installing dependencies from Pipfile.lock...
An error occured while installing!
Invalid requirement: 'crayons from git+https://github.com/kennethreitz/crayons.git@v0.1.2#egg=crayons==0.1.2'
It looks like a path. Does it exist ?


To activate this project's virtualenv, run the following:
 $ pipenv shell
```

These are the versions in my env:

```
$ pip freeze
appdirs==1.4.0
blindspin==2.0.0
click==6.7
click-completion==0.2.1
colorama==0.3.7
crayons==0.1.2
delegator.py==0.0.8
Jinja2==2.9.5
MarkupSafe==0.23
packaging==16.8
parse==1.6.6
pew==0.1.26
pexpect==4.2.1
pipenv==3.3.6
pipfile==0.0.1
ptyprocess==0.5.1
pyparsing==2.1.10
pythonz-bd==1.11.4
requests==2.13.0
requirements-parser==0.1.0
resumable-urlretrieve==0.1.4
six==1.10.0
toml==0.9.2
virtualenv==15.1.0
virtualenv-clone==0.2.6
```

Leaving out the ref attribute makes no difference.

Providing the `editable` flag leaves the package out of the lock file altogether:

```
$ cat Pipfile
[[source]]
url = "https://pypi.python.org/simple"
verify_ssl = true

[packages]
crayons = { git="https://github.com/kennethreitz/crayons.git", ref="v0.1.2", editable=1 }
```

produces:
```
$ cat Pipfile.lock
{
    "_meta": {
        "sources": [
            {
                "url": "https://pypi.python.org/simple",
                "verify_ssl": true
            }
        ],
        "requires": {},
        "hash": {
            "sha256": "fa6b6a1618451ef360028928588f70f6b3964f4dc8b0d371fb71a355b334abee"
        }
    },
    "develop": {},
    "default": {}
}
```

**Comment by nateprewitt** [comment]
Thanks for opening this @evertlammerts, we realized last week that this was regressed in the 3.3.X releases of pipenv. I'm currently working on getting this functionality reintegrated, so bear with me while we get this tidied. 

Since we hadn't gotten around to making a formal issue for this, we'll use this one to track this going forward.

**Comment by evertlammerts** [comment]
Thanks for the update. What's the last known version to have worked? I've just tried 3.2.14 and that doesn't seem to either.

Let me know if you want some help on this, with a few pointers I may be able to come up with a fix?

**Comment by nateprewitt** [comment]
Hrmm, so you've actually encountered a variant of what we noted in #198. I had to backtrack all the way to 0.2.8 to find a version that handles this as expected which is before I started working on the project. While this will generate the expected Pipfile.lock output, it doesn't seem that pipenv was able to install from those specs at the time.

That leads me to believe this specific case may have *never* worked. We need to make some changes in `do_lock`, `get_downloads_info` and `ensure_proper_casing`. I'd be open to reviewing a patch fixing this but will need tests ensuring the change doesn't break other functionality. This particular portion of the code is hard to test right now so I'm working on refactoring some of it. Which will then allow us to implement this functionality more cleanly.

**Comment by kennethreitz** [comment]
i do believe that `egg` has always been a requirement.

**Comment by evertlammerts** [comment]
Sorry, don't think I understand. Are you saying I can't install from Git?

**Comment by nateprewitt** [comment]
@evertlammerts, you *should* be able to install from git, but the functionality is currently broken. The use of `ref` and `editable` hadn't been tested before, and I can't find a version of pipenv where it ever did work correctly. We are working towards getting this addressed in a future release.

**Comment by evertlammerts** [comment]
Got it, thanks!

**Comment by chenull** [comment]
Hi, thanks for this great tools! i just want to report another error when installing a package using git:
```
$ pipenv --version
pipenv, version 3.4.1`
```

`$ pipenv install git+https://github.com/django/django.git`

```shell
Installing git+https://github.com/django/django.git...
Collecting git+https://github.com/django/django.git
  Cloning https://github.com/django/django.git to /private/var/folders/t_/cpyg93cx79zfjwgnp24fx7rw0000gn/T/pip-dfds2x7k-build
Requirement already satisfied: pytz in /Users/ayik/.virtualenvs/empu/lib/python3.6/site-packages/pytz-2016.10-py3.6.egg (from Django==2.0.dev20170214233504)
Installing collected packages: Django
  Found existing installation: Django 1.10.5
    Uninstalling Django-1.10.5:
      Successfully uninstalled Django-1.10.5
  Running setup.py install for Django: started
    Running setup.py install for Django: finished with status 'done'
Successfully installed Django-2.0.dev20170214233504

Adding git+https://github.com/django/django.git to Pipfile's [packages]...
Traceback (most recent call last):
  File "/Users/ayik/.local/bin/pipenv", line 11, in <module>
    sys.exit(cli())
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/cli.py", line 771, in install
    project.add_package_to_pipfile(package_name, dev)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/project.py", line 173, in add_package_to_pipfile
    self.write_toml(p)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/project.py", line 127, in write_toml
    f.write(format_toml(toml.dumps(data)))
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/toml.py", line 602, in dumps
    addtoretval, addtosections = _dump_sections(sections[section], section)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/toml.py", line 619, in _dump_sections
    if not re.match(r'^[A-Za-z0-9_-]+$', section):
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/re.py", line 172, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object
```

and... my `Pipfile` is now an empty file ! 😱

**Comment by nateprewitt** [comment]
Hey @chenull, this is a separate but definitely important issue. Can you please open this in another ticket so we can track it appropriately.

**Comment by chenull** [comment]
submitted new issue in #225

**Comment by zackkitzmiller** [comment]
I believe that is actually a dupe of https://github.com/kennethreitz/pipenv/issues/47

**Comment by nateprewitt** [comment]
Thanks @zackkitzmiller, this issue (#204) is definitely separate from #47. #225 is tangential to #47 but has some other nuances (files being erased). I'm leaving both of these open for the moment because they aren't fully resolved.

**Comment by zackkitzmiller** [comment]
@nateprewitt Sounds good. I'll just runoff then.

**Comment by simnalamburt** [comment]
I'm going to fix this issue. Is there anyone else who's currently working on this issue?

**Comment by nateprewitt** [comment]
I'm in the middle of working on this @simnalamburt. The problems for this are actually fairly systemic and I haven't fully synced up with Kenneth on finalizing some approaches.

The general overview is as follows:

We currently try to download every package name in the Pipfile from pip. This means a vcs dependency that exists in Pip gets overwritten. Separating that out requires a change in how we internally model the Pipfile without breaking our public interface for it.

We also need to have this similar approach for Pipfile.lock because when we go to install dependencies they are all required to have hashes in our current approach. I've rewritten this to do installs from two separate requirements files, one containing hashes for pip and the other installing from the vcs.

This is all mostly done but needs to be centralized into a clearer interface in projects.py which I haven't had a moment to sit down with for the last couple days due to everything else going on with this project and others.

If you'd like to peruse through the code and see if you have any further ideas, I'd be happy to read over them.

**Comment by simnalamburt** [comment]
Oh I didn't know that it's that complicated issue. I just test it on my own and I found that it's not that simple. But glad to hear that it's all mostly done! I'll give you a hand if you need any help to resolve this issue. Thanks!

**Comment by astynax** [comment]
Definitions like this
```
[packages]
crayons = { git="https://github.com/kennethreitz/crayons.git", ref="v0.1.2" }
```
aren't valid in TOML.

Pipenv will install from git, if you define Git-dependency as TOML [table](https://github.com/toml-lang/toml#table):
```
[packages]
foo = "*"

[packages.crayons]
git = "https://github.com/kennethreitz/crayons.git"
ref = "v0.1.2"
```

**Comment by nateprewitt** [comment]
@astynax, inline tables are very much so [valid TOML](https://github.com/toml-lang/toml#user-content-inline-table).

> Pipenv will install from git, if you define Git-dependency as TOML table

Are you suggesting that you're currently able to install from git with pipenv when using a table?

**Comment by astynax** [comment]
@nateprewitt Hmm... you are right about inline tables.
But `pipenv install git+...` still fails. And Pipfile with tables + `pipenv install` work for me.

**Comment by nateprewitt** [comment]
> But pipenv install git+... still fails.

Installing from `git` should still work fine in the form `pipenv install git+https://github.com/django/dango.git#egg=Django`. I believe this is noted farther up in the issue.

The problem we're discussing here is installing those dependencies again from the Pipfile/Pipfile.lock. There's a pretty detailed explanation of the current state of things two comments above your original.

The only reason that `crayons` works in this case is it's a dependency of `pipenv` itself and *MUST* be installed and you have chosen the same version number that `pip` will install by default. If you try another package that isn't hosted on PyPI such as the example below, you'll find your table syntax doesn't work.

```
[packages.coinbase]
git = "https://github.com/coinbase/coinbase-python"
version = "2.0.2"
```

**Comment by astynax** [comment]
@nateprewitt I mentioned the `crayons` in my comments just for example. I used toml-tables with `pipenv install` for another package (pylint) and installation was succeed.

**Comment by nateprewitt** [comment]
This should be resolved with #242 and released in 3.5.0.
